### PR TITLE
Move and test _apply_to_spatial_axes

### PR DESCRIPTION
**Pull request by jgersak**

### CODE REVIEW

File: core/eolearn/tests/test_utils/test_common.py [hunk 1]
```diff
@@ -50,3 +53,40 @@ def test_is_discrete_type(number_type, is_discrete):
...
+
+
+@pytest.mark.parametrize(
+    "map_function, data, spatial_axes, expected",
+    [
+        (partial(np.resize, new_shape=(2, 2)), np.zeros(shape=0), (1, 2), np.zeros((2, 2))),
+        (partial(np.resize, new_shape=(2, 2)), np.ones((2, 3, 4, 1)), (1, 2), np.ones((2, 2, 2, 1))),
+        (partial(np.resize, new_shape=(3, 2)), np.ones((2, 3, 4, 1)), (0, 1), np.ones((3, 2, 4, 1))),
+        (partial(np.resize, new_shape=(2, 3)), np.ones((2, 3, 4, 1)), (0, 2), np.ones((2, 3, 3, 1))),
+        (partial(np.resize, new_shape=(1, 2)), np.ones((2, 3, 4, 1)), (2, 3), np.ones((2, 3, 1, 2))),
```
Comment (zigaLuksic) [inline core/eolearn/tests/test_utils/test_common.py:65]: many of these are either redundant or hard to understand.
use a `new_shape` that is clearly different, such as `5, 6`.

File: core/eolearn/tests/test_utils/test_common.py [hunk 2]
```diff
@@ -50,3 +53,40 @@ def test_is_discrete_type(number_type, is_discrete):
...
+            partial(np.flip, axis=0),
+            np.arange(2 * 3 * 4).reshape((2, 3, 4, 1)),
+            (1, 2),
+            np.array(
+                [
+                    [[[8], [9], [10], [11]], [[4], [5], [6], [7]], [[0], [1], [2], [3]]],
+                    [[[20], [21], [22], [23]], [[16], [17], [18], [19]], [[12], [13], [14], [15]]],
+                ]
+            ),
+        ),
```
Comment (zigaLuksic) [inline core/eolearn/tests/test_utils/test_common.py:82]: i get lost in 'what is what' here frequently.
Find what is a common solution to this with `pytest`. Maybe `pytest.param` could help, maybe you have to extract the cases to a dataclass.

File: core/eolearn/tests/test_utils/test_common.py [hunk 3]
```diff
@@ -50,3 +53,40 @@ def test_is_discrete_type(number_type, is_discrete):
...
+        (lambda x: x + 1, np.arange(24).reshape((2, 3, 4, 1)), (1, 2), np.arange(1, 25).reshape((2, 3, 4, 1))),
+    ],
+)
+def test_apply_to_spatial_axes(
```
Comment (zigaLuksic) [inline core/eolearn/tests/test_utils/test_common.py:86]: what is the behaviour when the axes are reversed, e.g. `spatial_axes=(5,3)`. What if one or both of the axes are larger than `ndim`. You already have a 1-dim case that gets miraculously transformed to a 2-dim case, is that what we want?

File: features/eolearn/features/interpolation.py [hunk 4]
```diff
@@ -319,9 +319,11 @@ def interpolate_data(self, data: np.ndarray, times: np.ndarray, resampled_times:
...
             [
-                np.abs(resampled_times - o).argmin()
-                if np.min(resampled_times) <= o <= np.max(resampled_times)
-                else None
+                (
+                    np.abs(resampled_times - o).argmin()
+                    if np.min(resampled_times) <= o <= np.max(resampled_times)
+                    else None
+                )
                 for o in times
```
Comment (zigaLuksic) [inline features/eolearn/features/interpolation.py:327]: not relevant

File: core/eolearn/tests/test_utils/test_common.py (hunk 3 above)
Comment (jgersak) [reply core/eolearn/tests/test_utils/test_common.py:86]: spatial_axes=(5,3) its fine is same as spatial_axes=(3,5) if this is fine let say function is np.flip axis =0 is going to be 3 in both cases

File: core/eolearn/tests/test_utils/test_common.py (hunk 1 above)
Comment (zigaLuksic) [reply core/eolearn/tests/test_utils/test_common.py:65]: not axis, `new shape`

File: features/eolearn/features/interpolation.py (hunk 4 above)
Comment (jgersak) [reply features/eolearn/features/interpolation.py]: Revert.

File: core/eolearn/tests/test_utils/test_common.py (hunk 2 above)
Comment (jgersak) [reply core/eolearn/tests/test_utils/test_common.py]: Pytest.param applied hope it is less confusing.

File: core/eolearn/tests/test_utils/test_common.py (hunk 1 above)
Comment (jgersak) [reply core/eolearn/tests/test_utils/test_common.py]: test case updated

File: core/eolearn/tests/test_utils/test_common.py (hunk 3 above)
Comment (jgersak) [reply core/eolearn/tests/test_utils/test_common.py]: Add test cases, issue opened.

File: core/eolearn/tests/test_utils/test_common.py (hunk 2 above)
Comment (jgersak) [reply core/eolearn/tests/test_utils/test_common.py]: Extract the cases to a dataclass.

File: core/eolearn/tests/test_utils/test_common.py (hunk 1 above)
Comment (jgersak) [reply core/eolearn/tests/test_utils/test_common.py]: Updated in some cases, still want to test shrinking.

File: core/eolearn/core/utils/common.py [hunk 5]
```diff
@@ -102,6 +102,12 @@ def _apply_to_spatial_axes(
     After achieving 2D data slices the mapping function is applied. The data is then reconstructed into original form.
     """
 
+    if len(spatial_axes) != 2:
+        raise ValueError("Parameter `spatial_axes` should be tuple of two increasing int smaller then `data.ndim`")
```
Comment (zigaLuksic) [inline core/eolearn/core/utils/common.py:106]: not really required IMO, the type is specific enough

File: core/eolearn/core/utils/common.py [hunk 6]
```diff
@@ -102,6 +102,12 @@ def _apply_to_spatial_axes(
...
+    ax1, ax2 = spatial_axes
+    if ax1 >= ax2 or ax2 >= data.ndim:
+        raise ValueError("Parameter `spatial_axes` should be tuple of two increasing int smaller then `data.ndim`")
```
Comment (zigaLuksic) [inline core/eolearn/core/utils/common.py:109]: if you split this into two, you can provide nicer error messages.

File: core/eolearn/tests/test_utils/test_common.py [hunk 7]
```diff
@@ -61,16 +61,10 @@ class ApplyToTestCase:
     function: Callable[[np.ndarray], np.ndarray]
     data: np.ndarray
     spatial_axes: Tuple[int, int]
-    expected: np.ndarray
+    expected: Optional[np.ndarray] = None
 
 
 APPLY_TO_TEST_CASES = [
-    ApplyToTestCase(
```
Comment (zigaLuksic) [inline core/eolearn/tests/test_utils/test_common.py]: I would change the name of the dataclass to `ApplyToAxesTestCase`, currently it looks like you're applying stuff to tests

Comment (jgersak) [reply core/eolearn/tests/test_utils/test_common.py]: Agree changed.

File: core/eolearn/core/utils/common.py (hunk 6 above)
Comment (jgersak) [reply core/eolearn/core/utils/common.py]: Updated.

File: core/eolearn/core/utils/common.py (hunk 5 above)
Comment (jgersak) [reply core/eolearn/core/utils/common.py]: Removed.

File: core/eolearn/core/utils/common.py [hunk 8]
```diff
@@ -88,3 +88,34 @@ def generate_uid(prefix: str) -> str:
...
+    Recursively slices data into smaller-dimensional ones, until only the spatial axes remain. The indices of spatial
+    axes have to be adjusted if the recursion-axis is smaller than either one, e.g. spatial axes (1, 2) become (0, 1)
+    after splitting the 3D data along axis 0 into 2D arrays.
+
+    After achieving 2D data slices the mapping function is applied. The data is then reconstructed into original form.
+    """
+
+    ax1, ax2 = spatial_axes
+    if ax1 >= ax2:
+        raise ValueError("Values in `spatial_axes` should be increasing.")
```
Comment (zigaLuksic) [inline core/eolearn/core/utils/common.py:107]:
```suggestion
        raise ValueError(f"For parameter `spatial_axes` the second axis must be greater than first, got {spatial_axes}.")
```

File: core/eolearn/core/utils/common.py [hunk 9]
```diff
@@ -88,3 +88,34 @@ def generate_uid(prefix: str) -> str:
...
+
+    if ax2 >= data.ndim:
+        raise ValueError("Values in `spatial_axes` should smaller then `data.ndim`.")
```
Comment (zigaLuksic) [inline core/eolearn/core/utils/common.py:110]:
```suggestion
        raise ValueError(f"Values in `spatial_axes` must be smaller than `data.ndim`, got {spatial_axes} for data of dimension {data.ndim}.")
```

Black will surely complain about this one :see_no_evil:

File: core/eolearn/tests/test_utils/test_common.py [hunk 10]
```diff
@@ -50,3 +54,97 @@ def test_is_discrete_type(number_type, is_discrete):
...
+    ),
+    ApplyToAxesTestCase(
+        function=partial(np.resize, new_shape=(5, 6)),
+        data=np.ones((2, 3, 4, 1)),
+        spatial_axes=(0, 2),
+        expected=np.ones((5, 3, 6, 1)),
+    ),
+    ApplyToAxesTestCase(
+        function=partial(np.resize, new_shape=(2, 4)),
+        data=np.ones((2, 3, 4, 1)),
```
Comment (zigaLuksic) [inline core/eolearn/tests/test_utils/test_common.py:82]: just noticed that all data has ndim=4, adjust some test cases

Comment (jgersak) [reply core/eolearn/tests/test_utils/test_common.py:88]: add 2, 3 and 5 dim cases

**Review by zigaLuksic — APPROVED** [review]

**Pull request merged**